--- src/libcore/intrinsics.rs
+++ src/libcore/intrinsics.rs
@@ -678,5 +678,9 @@
     pub fn min_align_of_val<T: ?Sized>(_: &T) -> usize;
 
+    /// Obtain the length of a slice pointer
+    #[cfg(rust_compiler="mrustc")]
+    pub fn mrustc_slice_len<T>(pointer: *const [T]) -> usize;
+
     /// Gets a static string slice containing the name of a type.
     pub fn type_name<T: ?Sized>() -> &'static str;

--- src/libcore/slice/mod.rs
+++ src/libcore/slice/mod.rs
@@ -413,6 +413,8 @@
     #[inline]
     fn len(&self) -> usize {
-        unsafe {
-            mem::transmute::<&[T], Repr<T>>(self).len
-        }
+        #[cfg(not(rust_compiler="mrustc"))]
+        let rv = unsafe { mem::transmute::<&[T], Repr<T>>(self).len };
+        #[cfg(rust_compiler="mrustc")]
+        let rv = unsafe { ::intrinsics::mrustc_slice_len(self) };
+        rv
     }
--- src/llvm/cmake/modules/HandleLLVMOptions.cmake.orig	2017-06-27 11:37:39 UTC
+++ src/llvm/cmake/modules/HandleLLVMOptions.cmake
@@ -110,6 +110,7 @@ if(NOT (${TARGET_TRIPLE} MATCHES "darwin
         ${TARGET_TRIPLE} MATCHES "mingw" OR
         ${TARGET_TRIPLE} MATCHES "freebsd" OR
         ${TARGET_TRIPLE} MATCHES "netbsd" OR
+        ${TARGET_TRIPLE} MATCHES "dragonfly" OR
         ${TARGET_TRIPLE} MATCHES "openbsd") AND
    NOT LLVM_USE_SANITIZER)
   set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,-z,defs")
